br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -494,6 +502,10 @@ func (m *DupeDetector) RecordDataConflictError(ctx context.Context, stream DupKV
 			}
 			dataConflictInfos = dataConflictInfos[:0]
 		}
+
+		if algorithm == config.DupeResAlgErr {
+			return errors.Trace(common.ErrFoundDataConflictRecords.FastGenByArgs(m.tbl.Meta().Name, h.String(), m.decoder.DecodeRawRowDataAsStr(h, val)))
```

Comment:
This is the usage of DDL

https://github.com/pingcap/tidb/blob/b2c3e67fdca6bf641d92ed63b41d8aee4ccf374d/pkg/ddl/ingest/backend.go#L139-L152

Seems when it wants to fail fast for duplicate error, CollectRemoteDuplicateRows should not return error. Instead, it should return a `true` `hasDupe`, and ddl logic will then convert it to tikv.ErrKeyExists so TestAddIndexRemoteDuplicateCheck will pass.

This fail fast behaviour is different with import's fail fast behaviour, where it expects there's a non-nil error.

Do you think it's better that we change DupeController.CollectRemoteDuplicateRows to return 3 variables, `hasDupe bool, errorFromDupeResAlgErr error, runningErr error`?

Reply:
/cc @tangenta 

And in old ddl logic, DDL will not report the duplicate entry in error message, can you guide us how to return a correct format to build the error message?

Reply:
DDL behavior has been maintained to conform to the original behavior.